src/main/java/com/intellij/rt/debugger/agent/ConditionalBreakpointTransformer.java

```diff
@@ -0,0 +1,367 @@
+                                    throw new InstrumentationBpExceptionWrapper(e, instrumentationId);
+                                }
+                            }
+                        };
+                    }
+                }, 0, true);
+            } catch (Throwable e) {
+                if (e instanceof InstrumentationBpExceptionWrapper) {
+                    instrumentationFailed(e.getCause(), ((InstrumentationBpExceptionWrapper)e).instrumentationId);
+                } else {
+                    instrumentationFailed(e, -1);
+                }
+            }
+            return null;
+        }
+    }
+
+    private static Map<Integer, InstrumentationBreakpointMappingInfo> collectArgumentMapping(MethodNode method, Map<Integer, InstrumentationBreakpointInfo> lineNumbers) {
+        final Map<Integer, InstrumentationBreakpointMappingInfo> remappingInfo = new HashMap<>();
+
+        Set<Integer> visitedLineNumbers = new HashSet<>();
+        for (int instructionIndex = 0; instructionIndex < method.instructions.size(); instructionIndex++) {
+            AbstractInsnNode instruction = method.instructions.get(instructionIndex);
+            if (!(instruction instanceof LineNumberNode)) {
+                continue;
+            }
+
+            int lineNumber = ((LineNumberNode)instruction).line;
+            InstrumentationBreakpointInfo instrumentationBreakpointInfo = lineNumbers.get(lineNumber);
+            if (instrumentationBreakpointInfo != null) {
+                // skip non-trivial cases for now
+                if (visitedLineNumbers.contains(lineNumber)) {
+                    impossibleToInstrument("Several instructions marked with the same line " + lineNumber, extractIdFromFragmentClassName(instrumentationBreakpointInfo.fragmentClassName));
+                    remappingInfo.remove(lineNumber);
+                }
+                else {
+                    visitedLineNumbers.add(lineNumber);
+
+                    List<ArgumentGetter> argumentLoadGenerators = new ArrayList<>();
+                    for (String argumentName : instrumentationBreakpointInfo.argumentNames) {
+                        if (argumentName.equals("this")/* || argumentName.startsWith("this$")*/) {
+                            argumentLoadGenerators.add(new ArgumentGetter() {
+                                @Override
+                                public void generateArgumentLoad(MethodVisitor mv) {
+                                    mv.visitVarInsn(Opcodes.ALOAD, 0);
+                                }
+                            });
+                        }
+                        else {
+                            boolean isFound = false;
+                            for (LocalVariableNode localVariableNode : method.localVariables) {
+                                if (localVariableNode.name.equals(argumentName) &&
+                                        method.instructions.indexOf(localVariableNode.start) <= instructionIndex &&
```

Comment:
I still don't like this approach with loading of the whole class into `ClassNode` tree. All this stuff could be done more efficiently using visitors. Is it your intentional choice or just a prototype?